### PR TITLE
Added establishment (GIAS) refresh job

### PR DESCRIPTION
**Pull request by hortha**
### Context

As part of the workforce data import process we need to map from the format establishment details are stored in TPS to known establishments in GIAS.

### Changes proposed in this pull request

Create a job which runs daily to import the minimal fields we need from GIAS for the workforce data establishment mapping + data which may be needed to provide via the TRS API

### Guidance to review

Database change + background job + tests

### Checklist

-   [x] Attach to Trello card
-   [x] Rebased master
-   [x] Cleaned commit history
-   [x] Tested by running locally

### CODE REVIEW

File: TeachingRecordSystem/src/TeachingRecordSystem.Core/Jobs/HostApplicationBuilderExtensions.cs [hunk 1]
```diff
@@ -101,6 +102,15 @@ public static IHostApplicationBuilder AddBackgroundJobs(this IHostApplicationBui
                     job => job.Execute(),
                     Cron.Never);
 
+                var giasOptions = sp.GetService<IOptions<GiasOptions>>();
+                if (giasOptions != null)
```
Comment (gunndabad) [inline TeachingRecordSystem/src/TeachingRecordSystem.Core/Jobs/HostApplicationBuilderExtensions.cs:106]: Do we need this null check?

File: TeachingRecordSystem/src/TeachingRecordSystem.Core/Jobs/HostApplicationBuilderExtensions.cs [hunk 2]
```diff
@@ -101,6 +102,15 @@ public static IHostApplicationBuilder AddBackgroundJobs(this IHostApplicationBui
...
+                {
+                    recurringJobManager.AddOrUpdate<RefreshEstablishmentsJob>(
+                    nameof(RefreshEstablishmentsJob),
+                    job => job.ExecuteAsync(CancellationToken.None),
+                    giasOptions.Value.RefreshEstablishmentsJobSchedule);
```
Comment (gunndabad) [inline TeachingRecordSystem/src/TeachingRecordSystem.Core/Jobs/HostApplicationBuilderExtensions.cs:111]: Nit:
```suggestion
                    recurringJobManager.AddOrUpdate<RefreshEstablishmentsJob>(
                        nameof(RefreshEstablishmentsJob),
                        job => job.ExecuteAsync(CancellationToken.None),
                        giasOptions.Value.RefreshEstablishmentsJobSchedule);
```

File: TeachingRecordSystem/src/TeachingRecordSystem.Core/Services/Establishments/EstablishmentCsvRowMinimum.cs [hunk 3]
```diff
@@ -0,0 +1,64 @@
+using CsvHelper.Configuration.Attributes;
+
+namespace TeachingRecordSystem.Core.Services.Establishments;
+
+public class EstablishmentCsvRowMinimum
```
Comment (gunndabad) [inline TeachingRecordSystem/src/TeachingRecordSystem.Core/Services/Establishments/EstablishmentCsvRowMinimum.cs:5]:
```suggestion
public record EstablishmentCsvRow
```

File: TeachingRecordSystem/src/TeachingRecordSystem.Core/Services/Establishments/EstablishmentCsvRowMinimum.cs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+{
+    [Name("URN")]
+    public required string Urn { get; init; }
```
Comment (gunndabad) [inline TeachingRecordSystem/src/TeachingRecordSystem.Core/Services/Establishments/EstablishmentCsvRowMinimum.cs:8]:
```suggestion
    public required int Urn { get; init; }
```

File: TeachingRecordSystem/src/TeachingRecordSystem.Core/Jobs/HostApplicationBuilderExtensions.cs (hunk 1 above)
Comment (hortha) [reply TeachingRecordSystem/src/TeachingRecordSystem.Core/Jobs/HostApplicationBuilderExtensions.cs:106]: that's what I'm using to not add the job

File: TeachingRecordSystem/src/TeachingRecordSystem.Core/Models/Establishment.cs [hunk 5]
```diff
@@ -0,0 +1,22 @@
+namespace TeachingRecordSystem.Core.Models;
+
+public class Establishment
+{
+    public required int Urn { get; init; }
```
Comment (gunndabad) [inline TeachingRecordSystem/src/TeachingRecordSystem.Core/Models/Establishment.cs:5]: I think we should make all these properties `required` and `get; init;`

File: TeachingRecordSystem/src/TeachingRecordSystem.Core/DataStore/Postgres/Models/Establishment.cs [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    public required int EstablishmentTypeGroupCode { get; set; }
+    public required string EstablishmentTypeGroupName { get; set; }
+    public required int EstablishmentStatusCode { get; set; }
+    public required string EstablishmentStatusName { get; set; }
+    public string? Street { get; set; }
+    public string? Locality { get; set; }
+    public string? Address3 { get; set; }
+    public string? Town { get; set; }
+    public string? County { get; set; }
+    public string? Postcode { get; set; }
```
Comment (gunndabad) [inline TeachingRecordSystem/src/TeachingRecordSystem.Core/DataStore/Postgres/Models/Establishment.cs:25]:
```suggestion
    public required string? Street { get; set; }
    public required string? Locality { get; set; }
    public required string? Address3 { get; set; }
    public required string? Town { get; set; }
    public required string? County { get; set; }
    public required string? Postcode { get; set; }
```

**Comment by hortha** [comment]
[Import GIAS data in TRS](https://trello.com/c/2IS72EaJ/1806-import-gias-data-in-trs)

File: TeachingRecordSystem/src/TeachingRecordSystem.Core/Jobs/HostApplicationBuilderExtensions.cs [hunk 7]
```diff
@@ -101,6 +102,12 @@ public static IHostApplicationBuilder AddBackgroundJobs(this IHostApplicationBui
                     job => job.Execute(),
                     Cron.Never);
 
+                var giasOptions = sp.GetService<IOptions<GiasOptions>>();
```
Comment (gunndabad) [inline TeachingRecordSystem/src/TeachingRecordSystem.Core/Jobs/HostApplicationBuilderExtensions.cs:105]: Nit:
```suggestion
                var giasOptions = sp.GetRequiredService<IOptions<GiasOptions>>();
```

File: TeachingRecordSystem/src/TeachingRecordSystem.Worker/appsettings.Development.json [hunk 8]
```diff
@@ -2,6 +2,10 @@
   "DqtReporting": {
     "ProcessAllEntityTypesConcurrently": false
   },
+  "Gias": {
+    "BaseDownloadAddress": "https://ea-edubase-api-prod.azurewebsites.net/edubase/downloads/public/",
```
Comment (gunndabad) [inline TeachingRecordSystem/src/TeachingRecordSystem.Worker/appsettings.Development.json:6]: Don't need this line if it's on the root `appsettings.json`

**Review by gunndabad — APPROVED** [review]

**Pull request merged**